File: meshcall-ts/src/router.ts
import { spawn, type ChildProcessWithoutNullStreams } from "node:child_process";
import { accessSync, constants, statSync } from "node:fs";
import { delimiter, dirname, resolve } from "node:path";
import { fileURLToPath } from "node:url";

import { unixSocketPath } from "./endpoint.js";
import { PROTOCOL_VERSION } from "./protocol.js";

export interface WebSocketRouterOptions {
  readonly host?: string;
  readonly port?: number;
  readonly unixPath?: string;
  readonly maxFrameSize?: number;
  readonly binaryPath?: string;
  readonly authFile?: string;
  readonly startupTimeoutMs?: number;
  readonly shutdownTimeoutMs?: number;
}

function executable(path: string): boolean {
  try {
    accessSync(path, constants.X_OK);
    return statSync(path).isFile();
  } catch { return false; }
}

function routerBinary(configured?: string): string {
  const selected = configured ?? process.env.MESHCALL_ROUTER_BINARY;
  if (selected !== undefined) {
    const path = resolve(selected);
    if (!executable(path)) throw new Error("Router binary is not executable: " + path);
    return path;
  }
  const filename = process.platform === "win32" ? "meshcall-router.exe" : "meshcall-router";
  const directory = dirname(fileURLToPath(import.meta.url));
  const candidates = [
    resolve(directory, "../../bin", filename),
    resolve(directory, "../../../meshcall-router/bin", filename),
    ...(process.env.PATH ?? "").split(delimiter).filter(Boolean).map((path) => resolve(path, filename)),
  ];
  const found = candidates.find(executable);
  if (found !== undefined) return found;
  throw new Error("Go Router binary not found. Build meshcall-router or set MESHCALL_ROUTER_BINARY / binaryPath.");
}

async function beforeTimeout(promise: Promise<unknown>, ms: number): Promise<boolean> {
  let timer: ReturnType<typeof setTimeout> | undefined;
  try {
    return await Promise.race([
      promise.then(() => true),
      new Promise<false>((resolve) => { timer = setTimeout(() => resolve(false), ms); }),
    ]);
  } finally { if (timer !== undefined) clearTimeout(timer); }
}

/** Owns a Go child process. No registration, authentication or routing runs here. */
export class WebSocketRouter {
  private child: ChildProcessWithoutNullStreams | undefined;
  private exited: Promise<number | null> | undefined;
  private lifecycle: Promise<void> = Promise.resolve();
  private running = false;
  private port: number | undefined;
  private diagnostic = "";
  private readonly unixPath: string | undefined;
  private readonly startupTimeout: number;
  private readonly shutdownTimeout: number;

  public constructor(private readonly options: WebSocketRouterOptions = {}) {
    if (options.unixPath !== undefined && (options.host !== undefined || options.port !== undefined)) {
      throw new TypeError("Unix socket cannot be combined with host or port");
    }
    if (options.port !== undefined && (!Number.isSafeInteger(options.port) || options.port < 0 || options.port > 65535)) {
      throw new RangeError("Port must be between 0 and 65535");
    }
    const maxFrameSize = options.maxFrameSize ?? 1024 * 1024;
    if (!Number.isSafeInteger(maxFrameSize) || maxFrameSize < 1 || maxFrameSize > 64 * 1024 * 1024) {
      throw new RangeError("maxFrameSize must be between 1 and 67108864");
    }
    this.startupTimeout = options.startupTimeoutMs ?? 10000;
    this.shutdownTimeout = options.shutdownTimeoutMs ?? 5000;
    for (const timeout of [this.startupTimeout, this.shutdownTimeout]) {
      if (!Number.isSafeInteger(timeout) || timeout <= 0) throw new RangeError("Process timeouts must be positive integers");
    }
    this.unixPath = options.unixPath === undefined ? undefined : unixSocketPath(options.unixPath);
  }

  public get isRunning(): boolean {
    return this.running && this.child !== undefined && this.child.exitCode === null && this.child.signalCode === null;
  }
  public get pid(): number | undefined { return this.isRunning ? this.child?.pid : undefined; }
  public get boundPort(): number {
    if (this.unixPath !== undefined) throw new Error("Unix socket listener does not have a TCP port");
    if (!this.isRunning || this.port === undefined) throw new Error("Go Router is not running");
    return this.port;
  }

  public start(): Promise<void> {
    return this.serialize(async () => {
      if (this.isRunning) return;
      await this.stopProcess();
      this.diagnostic = "";
      const args = ["--shutdown-on-stdin-close", "--max-frame-size", String(this.options.maxFrameSize ?? 1024 * 1024)];
      if (this.unixPath === undefined) args.push("--host", this.options.host ?? "127.0.0.1", "--port", String(this.options.port ?? 0));
      else args.push("--unix-socket", this.unixPath);
      if (this.options.authFile !== undefined) args.push("--auth-file", resolve(this.options.authFile));
      const child = spawn(routerBinary(this.options.binaryPath), args, { stdio: "pipe", windowsHide: true });
      this.child = child;
      child.stdin.on("error", () => undefined);
      this.exited = new Promise((resolve) => {
        child.once("close", (code) => {
          if (this.child === child) this.running = false;
          resolve(code);
        });
      });
      child.stderr.on("data", (chunk: Buffer) => {
        this.diagnostic = (this.diagnostic + chunk.toString("utf8")).slice(-8192);
      });
      let timer: ReturnType<typeof setTimeout> | undefined;
      try {
        await new Promise<void>((resolve, reject) => {
          let buffer = Buffer.alloc(0);
          let received = false;
          timer = setTimeout(() => reject(new Error("Go Router readiness timed out")), this.startupTimeout);
          child.on("error", reject);
          child.once("close", () => reject(new Error("Go Router exited before readiness")));
          child.stdout.on("data", (chunk: Buffer) => {
            if (received) return; // Continue draining stdout without retaining it.
            buffer = Buffer.concat([buffer, chunk]);
            if (buffer.length > 16384) { reject(new Error("Invalid Go Router readiness message")); return; }
            const newline = buffer.indexOf(10);
            if (newline < 0) return;
            received = true;
            try {
              const ready = JSON.parse(buffer.subarray(0, newline).toString("utf8")) as Record<string, unknown>;
              if (ready.kind !== "router.ready" || ready.protocol !== PROTOCOL_VERSION || ready.pid !== child.pid) {
                throw new Error("Invalid Go Router readiness message");
              }
              if (this.unixPath === undefined) {
                if (typeof ready.port !== "number" || !Number.isSafeInteger(ready.port) || ready.port < 1 || ready.port > 65535) {
                  throw new Error("Invalid Go Router bound port");
                }
                this.port = ready.port;
              } else if (ready.unix_path !== this.unixPath) throw new Error("Invalid Go Router Unix socket path");
              resolve();
            } catch (error) { reject(error); }
          });
        });
        if (child.exitCode !== null || child.signalCode !== null) throw new Error("Go Router exited during startup");
        this.running = true;
      } catch (error) {
        await this.stopProcess();
        throw new Error("Could not start Go Router: " + String(error) +
          (this.diagnostic ? "\n" + this.diagnostic.trim() : ""), { cause: error });
      } finally { if (timer !== undefined) clearTimeout(timer); }
    });
  }

  public stop(): Promise<void> { return this.serialize(() => this.stopProcess()); }
  public close(): Promise<void> { return this.stop(); }
  public waitClosed(): Promise<number | null> {
    if (this.exited === undefined) return Promise.reject(new Error("Go Router is not running"));
    return this.exited;
  }

  private serialize(operation: () => Promise<void>): Promise<void> {
    const result = this.lifecycle.then(operation);
    this.lifecycle = result.catch(() => undefined);
    return result;
  }

  private async stopProcess(): Promise<void> {
    const child = this.child;
    const exited = this.exited;
    this.running = false;
    this.port = undefined;
    if (child !== undefined && exited !== undefined) {
      child.stdin.end();
      if (!await beforeTimeout(exited, this.shutdownTimeout)) {
        child.kill("SIGTERM");
        if (!await beforeTimeout(exited, this.shutdownTimeout)) {
          child.kill("SIGKILL");
          await exited;
        }
      }
    }
    this.child = undefined;
    this.exited = undefined;
  }
}

File: meshcall-py/src/meshcall/router.py
from __future__ import annotations

import asyncio
import contextlib
import json
import os
import shutil
from pathlib import Path
from typing import Self

from meshcall.errors import DriverStateError
from meshcall.protocol import PROTOCOL_VERSION


def _router_binary(binary_path: str | Path | None) -> str:
    executable = "meshcall-router.exe" if os.name == "nt" else "meshcall-router"
    configured = binary_path or os.environ.get("MESHCALL_ROUTER_BINARY")
    if configured:
        candidate = Path(configured).expanduser().absolute()
        if not candidate.is_file() or not os.access(candidate, os.X_OK):
            raise DriverStateError(f"Router binary is not executable: {candidate}")
        return str(candidate)
    candidates = [Path(__file__).parent / "bin" / executable]
    # A demo can install a wheel into its own venv rather than an editable copy.
    # Find the enclosing source checkout from either the package or the caller.
    roots = dict.fromkeys(
        [
            *Path(__file__).resolve().parents,
            Path.cwd(),
            *Path.cwd().parents,
        ]
    )
    candidates.extend(
        root / "meshcall-router" / "bin" / executable
        for root in roots
        if (root / "meshcall-router" / "go.mod").is_file()
    )
    for candidate in candidates:
        if candidate.is_file() and os.access(candidate, os.X_OK):
            return str(candidate)
    installed = shutil.which(executable)
    if installed:
        return installed
    raise DriverStateError(
        "Go Router binary not found. Build meshcall-router or set "
        "MESHCALL_ROUTER_BINARY / binary_path to a prebuilt executable."
    )


class WebSocketRouter:
    """Own a Go Router process; all RPC routing and authentication run in Go."""

    def __init__(
        self,
        host: str | None = None,
        port: int | None = None,
        *,
        unix_path: str | Path | None = None,
        max_frame_size: int = 1024 * 1024,
        binary_path: str | Path | None = None,
        auth_file: str | Path | None = None,
        startup_timeout: float = 10,
        shutdown_timeout: float = 5,
    ) -> None:
        if unix_path is not None and (host is not None or port is not None):
            raise ValueError("Unix socket cannot be combined with host or port")
        if port is not None and not 0 <= port <= 65535:
            raise ValueError("Port must be between 0 and 65535")
        if not 1 <= max_frame_size <= 64 * 1024 * 1024:
            raise ValueError("max_frame_size must be between 1 and 67108864")
        if startup_timeout <= 0 or shutdown_timeout <= 0:
            raise ValueError("Process timeouts must be positive")
        self.host = host or "127.0.0.1"
        self.port = 0 if port is None else port
        self.unix_path = Path(unix_path).absolute() if unix_path is not None else None
        if self.unix_path is not None and len(str(self.unix_path).encode()) > 103:
            raise DriverStateError(
                "Unix socket path exceeds the portable 103-byte limit"
            )
        self.max_frame_size = max_frame_size
        self.binary_path = binary_path
        self.auth_file = Path(auth_file).absolute() if auth_file is not None else None
        self.startup_timeout = startup_timeout
        self.shutdown_timeout = shutdown_timeout
        self._process: asyncio.subprocess.Process | None = None
        self._bound_port: int | None = None
        self._ready = False
        self._lock = asyncio.Lock()
        self._stderr = bytearray()
        self._readers: list[asyncio.Task[None]] = []

    @property
    def is_running(self) -> bool:
        return (
            self._ready
            and self._process is not None
            and self._process.returncode is None
        )

    @property
    def pid(self) -> int | None:
        return (
            self._process.pid if self.is_running and self._process is not None else None
        )

    @property
    def bound_port(self) -> int:
        if self.unix_path is not None:
            raise DriverStateError("Unix socket listener does not have a TCP port")
        if not self.is_running or self._bound_port is None:
            raise DriverStateError("Go Router is not running")
        return self._bound_port

    async def _drain(self, stream: asyncio.StreamReader, *, diagnostic: bool) -> None:
        while data := await stream.read(4096):
            if diagnostic:
                self._stderr.extend(data)
                del self._stderr[:-8192]

    async def start(self) -> None:
        async with self._lock:
            if self.is_running:
                return
            await self._stop()
            command = [
                _router_binary(self.binary_path),
                "--shutdown-on-stdin-close",
                "--max-frame-size",
                str(self.max_frame_size),
            ]
            if self.unix_path is None:
                command.extend(["--host", self.host, "--port", str(self.port)])
            else:
                command.extend(["--unix-socket", str(self.unix_path)])
            if self.auth_file is not None:
                command.extend(["--auth-file", str(self.auth_file)])
            self._stderr.clear()
            try:
                process = await asyncio.create_subprocess_exec(
                    *command,
                    stdin=asyncio.subprocess.PIPE,
                    stdout=asyncio.subprocess.PIPE,
                    stderr=asyncio.subprocess.PIPE,
                    limit=16384,
                )
                self._process = process
                assert process.stdout is not None and process.stderr is not None
                self._readers.append(
                    asyncio.create_task(self._drain(process.stderr, diagnostic=True))
                )
                line = await asyncio.wait_for(
                    process.stdout.readline(), timeout=self.startup_timeout
                )
                if not line:
                    raise DriverStateError("Go Router exited before readiness")
                ready = json.loads(line)
                if (
                    not isinstance(ready, dict)
                    or ready.get("kind") != "router.ready"
                    or ready.get("protocol") != PROTOCOL_VERSION
                    or ready.get("pid") != process.pid
                ):
                    raise DriverStateError("Invalid Go Router readiness message")
                if self.unix_path is None:
                    port = ready.get("port")
                    if type(port) is not int or not 1 <= port <= 65535:
                        raise DriverStateError("Invalid Go Router bound port")
                    self._bound_port = port
                elif ready.get("unix_path") != str(self.unix_path):
                    raise DriverStateError("Invalid Go Router Unix socket path")
                self._readers.append(
                    asyncio.create_task(self._drain(process.stdout, diagnostic=False))
                )
                if process.returncode is not None:
                    raise DriverStateError("Go Router exited during startup")
                self._ready = True
            except asyncio.CancelledError:
                await self._stop()
                raise
            except Exception as exc:
                await self._stop()
                detail = self._stderr.decode(errors="replace").strip()
                message = f"Could not start Go Router: {exc}"
                if detail:
                    message += f"\n{detail}"
                raise DriverStateError(message) from exc

    async def stop(self) -> None:
        async with self._lock:
            await self._stop()

    async def _stop(self) -> None:
        process, self._process = self._process, None
        self._ready = False
        self._bound_port = None
        if process is not None:
            if process.stdin is not None:
                process.stdin.close()
            if process.returncode is None:
                try:
                    await asyncio.wait_for(process.wait(), self.shutdown_timeout)
                except TimeoutError:
                    with contextlib.suppress(ProcessLookupError):
                        process.terminate()
                    try:
                        await asyncio.wait_for(process.wait(), self.shutdown_timeout)
                    except TimeoutError:
                        with contextlib.suppress(ProcessLookupError):
                            process.kill()
                        await process.wait()
            await process.wait()
        if self._readers:
            await asyncio.gather(*self._readers)
            self._readers.clear()

    async def wait_closed(self) -> int:
        if self._process is None:
            raise DriverStateError("Go Router is not running")
        return await self._process.wait()

    async def __aenter__(self) -> Self:
        await self.start()
        return self

    async def __aexit__(self, *_: object) -> None:
        await self.stop()
